src/detect-origin.c

```diff
@@ -0,0 +1,432 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#include "suricata-common.h"
+#include "assert.h"
+
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-origin.h"
+
+static int DetectOriginMatch(
+        DetectEngineThreadCtx *, Packet *, const Signature *, const SigMatchCtx *);
+static int DetectOriginSetup(DetectEngineCtx *, Signature *, const char *);
+static void DetectOriginFree(DetectEngineCtx *, void *);
+
+#ifdef UNITTESTS
+static void DetectOriginRegisterTests(void);
+#endif
+
+/**
+ * \brief Registration function for origin: trusted/untrusted
+ *
+ * Only supporting roles of trusted/untrusted for faster matching.
+ * The origin being trusted/untrusted refers to the originating
+ * packet of a flow and the role of the live device for that packet.
+ */
+void DetectOriginRegister(void)
+{
+    sigmatch_table[DETECT_ORIGIN].name = "origin";
+    sigmatch_table[DETECT_ORIGIN].desc = "Match the origin of the traffic based on the role "
+                                         "(trusted/untrusted) of the interface.";
+    sigmatch_table[DETECT_ORIGIN].Match = DetectOriginMatch;
+    sigmatch_table[DETECT_ORIGIN].Setup = DetectOriginSetup;
+    sigmatch_table[DETECT_ORIGIN].Free = DetectOriginFree;
+    sigmatch_table[DETECT_ORIGIN].flags |= SIGMATCH_IPONLY_COMPAT;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_ORIGIN].RegisterTests = DetectOriginRegisterTests;
+#endif
+}
+
+/**
+ * \brief match origin role against a packet.
+ *
+ * \param det_ctx pointer to the pattern matcher thread
+ * \param p pointer to the current packet
+ * \param m pointer to the sigmatch context that we will cast into DetectOriginData
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+int DetectOriginMatch(
+        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+{
+    int ret = 0;
+    DetectOriginData *originData = (DetectOriginData *)ctx;
+
+    if (originData == NULL)
+        return ret;
+
+    if (p->flow && p->flow->livedev) {
+        if (originData->role == p->flow->livedev->role) {
```

Comment:
For the origin keyword we want to match on from which side a session originated. To accomplish this we use the live device associated with the flow. If we match on the packet the role will flip back and forth within a session depending on the direction of the communication for a given packet.

Reply:
This can simply be done with `flow:from_client`?

Reply:
We can't just rely on flow:from_client because it's still not device specific. What we're trying to accomplish is to have the ability in IPS mode to alert or mitigate against traffic from a specific interface. 

We could have a client/server operating on either side, but lets say we're only concerned when that communication originates from an "untrusted" device. We don't want to block traffic originating from the "trusted" device.